Changes to build with MinGW [WIP]

t-hey opened the pull request:
## Description
The changes in this pull request is to be able to build QGIS with MinGW.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

t-hey committed:
Update main.cpp

Changed Q_OS_WIN to _MSC_VER

t-hey committed:
Update qgsapplication.cpp

Changed code to compile with MINGW

t-hey committed:
Update qgsfeaturerequest.h

Changed code to CORE_EXPORT the whole class.

t-hey committed:
Update qgsmaplayermodel.cpp

Added code to return if the layers list is empty.

NathanW2 commented:
Is there a reason to not use VS build tools?

On Mon., 20 Nov. 2017, 6:40 pm theuns heydenrych, <notifications@github.com>
wrote:

> Description
>
> The changes in this pull request is to be able to build QGIS with MinGW.
> Checklist
>
> Reviewing is a process done by project maintainers, mostly on a volunteer
> basis. We try to keep the overhead as small as possible and appreciate if
> you help us to do so by completing the following items. Feel free to ask in
> a comment if you have troubles with any of them.
>
>
>    - Commit messages are descriptive and explain the rationale for changes
>    - Commits which fix bugs include fixes #11111 in the commit message
>    next to the description
>    - Commits which add new features are tagged with [FEATURE] in the
>    commit message
>    - Commits which change the UI or existing user workflows are tagged
>    with [needs-docs] in the commit message and containt sufficient
>    information in the commit message to be documented
>    - I have read the QGIS Coding Standards
>    <https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html>
>    and this PR complies with them
>    - This PR passes all existing unit tests (test results will be
>    reported by travis-ci after opening this PR)
>    - New unit tests have been added for core changes
>    - I have run the scripts/prepare-commit.sh script
>    <https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis>
>    before each commit
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/qgis/QGIS/pull/5675
> Commit Summary
>
>    - Update main.cpp
>    - Update qgsapplication.cpp
>    - Update qgsfeaturerequest.h
>    - Update qgsmaplayermodel.cpp
>
> File Changes
>
>    - *M* src/app/main.cpp
>    <https://github.com/qgis/QGIS/pull/5675/files#diff-0> (2)
>    - *M* src/core/qgsapplication.cpp
>    <https://github.com/qgis/QGIS/pull/5675/files#diff-1> (2)
>    - *M* src/core/qgsfeaturerequest.h
>    <https://github.com/qgis/QGIS/pull/5675/files#diff-2> (16)
>    - *M* src/core/qgsmaplayermodel.cpp
>    <https://github.com/qgis/QGIS/pull/5675/files#diff-3> (5)
>
> Patch Links:
>
>    - https://github.com/qgis/QGIS/pull/5675.patch
>    - https://github.com/qgis/QGIS/pull/5675.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/5675>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3EWE18vPRN24DSct8dbTp25HxPbgks5s4TsGgaJpZM4Qj6Eg>
> .
>

t-hey committed:
Update qgsmaplayermodel.cpp

Fixed code layout

jef-n commented on src/core/qgsmaplayermodel.cpp:
```diff
@@ -154,6 +154,11 @@ void QgsMapLayerModel::removeLayers( const QStringList &layerIds )
 
 void QgsMapLayerModel::addLayers( const QList<QgsMapLayer *> &layers )
 {
+  if ( layers.size() == 0 )
```
isEmpty()

jef-n commented on src/core/qgsfeaturerequest.h:
```diff
@@ -218,47 +218,47 @@ class CORE_EXPORT QgsFeatureRequest
      * \since QGIS 2.14
      */
-    class OrderBy : public QList<QgsFeatureRequest::OrderByClause>
+    class CORE_EXPORT OrderBy : public QList<QgsFeatureRequest::OrderByClause>
```
does this build with MSVC?

t-hey replied:
@jef-n I unfortunately dont have a MSVC toolchain to test this.

nyalldawson replied:
I've given this a test, and unfortunately it does break the msvc build. Does mingw build if you leave individual member CORE_EXPORTs, but insert this new class CORE_EXPORT? If so we could just #ifdef two different declarations for the class. If not, I can't think of any other way to resolve this.

t-hey committed:
Update qgsmaplayermodel.cpp

Fixed the code using jef-n suggestion.

t-hey committed:
Update qgsmaplayermodel.cpp

Removed space.

NathanW2 commented:
What is the reason to not use MSVC?

On Tue., 21 Nov. 2017, 3:27 pm theuns heydenrych, <notifications@github.com>
wrote:

> *@t-hey* commented on this pull request.
> ------------------------------
>
> In src/core/qgsfeaturerequest.h
> <https://github.com/qgis/QGIS/pull/5675#discussion_r152182287>:
>
> > @@ -218,47 +218,47 @@ class CORE_EXPORT QgsFeatureRequest
>       *
>       * \since QGIS 2.14
>       */
> -    class OrderBy : public QList<QgsFeatureRequest::OrderByClause>
> +    class CORE_EXPORT OrderBy : public QList<QgsFeatureRequest::OrderByClause>
>
> @jef-n <https://github.com/jef-n> I unfortunately dont have a MSVC
> toolchain to test this.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/5675#discussion_r152182287>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3DCgP_lZz8fnyu2L7B4dAxEpTQFtks5s4l9IgaJpZM4Qj6Eg>
> .
>

whatnick commented:
Building with MSVC may not be feasible in certain environments where MSVC licensing combined with network access restrictions make it impossible to use that compiler. A working build in MingW will produce compatible binaries if someone is willing to maintain the build system.

NathanW2 commented:
I guess but you don't need full VS to build QGIS. Just the 2015 build tools
(which is free) which are just really cl.exe and the toolchain.   As long
as the VS builds don't break from the changes I'm ok, I just don't want to
see support for something added and then dropped later because we can't
maintain it.

On Thu, Nov 23, 2017 at 1:34 PM, Tisham Dhar <notifications@github.com>
wrote:

> Building with MSVC may not be feasible in certain environments where MSVC
> licensing combined with network access restrictions make it impossible to
> use that compiler. A working build in MingW will produce compatible
> binaries if someone is willing to maintain the build system.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/5675#issuecomment-346525690>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3JIozg-w8FfOery2y3SslMIguG_uks5s5OfMgaJpZM4Qj6Eg>
> .
>

nyalldawson commented on src/core/qgsmaplayermodel.cpp:
```diff
@@ -154,6 +154,11 @@ void QgsMapLayerModel::removeLayers( const QStringList &layerIds )
 
 void QgsMapLayerModel::addLayers( const QList<QgsMapLayer *> &layers )
 {
+  if ( layers.isEmpty() )
```
@t-hey 

what's this change intended to address? I'm guessing it's unrelated to the mingw build, but was it a result of a bug you hit?

t-hey replied:
Yes, Q_ASSERT triggers in the beginInsertRows statement in this procedure at startup.

At startup the input parameters for beginInsertRows, "first" is 0 and "last" is -1, then Q_ASSERT(last >= first) triggers in the [QAbstractItemModel](https://github.com/qt/qtbase/blob/5.9.3/src/corelib/itemmodels/qabstractitemmodel.cpp).

Also if mAllowEmpty was true at startup, it would by fine, and also the addition of "offset" should be only set on the "last" paramater of beginInsertRows statement. 

Then there is no need for checking isEmpty() on the incoming layers list.

t-hey committed:
Update qgsfeaturerequest.h

Replaced the CORE_EXPORT macro.
Changed the default constructor, to be implemented in the cpp file.

t-hey committed:
Merge pull request #1 from qgis/master

Merge form QGIS master

nyalldawson commented:
Superseded by #5799
